File: Solutions/derived/derived.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

#include "alloc2d.h"

#define N 10//# of rows
#define M 10//# of columns

int main( int argc, char **argv ) {
    int ierror;
    int i, j;
    int world_rank, world_size;

    ierror = MPI_Init( &argc, &argv );
    ierror = MPI_Comm_rank( MPI_COMM_WORLD, &world_rank );
    ierror = MPI_Comm_size( MPI_COMM_WORLD, &world_size );

    int **A = alloc2dInt( N, M );
   //Rank 0 has a matrix of all zeros and Rank 1 has a matrix of all 1s  
   for ( i = 0; i < N; i++ ) {
        for ( j = 0; j < M; j++ ) {
            A[i][j] = world_rank;
        }
    }

    MPI_Datatype matRow, matCol, subMat;

    MPI_Type_vector( 1, M, 1, MPI_INT, &matRow );
    MPI_Type_commit( &matRow );
    MPI_Type_vector( N, 1, M, MPI_INT, &matCol );
    MPI_Type_commit( &matCol );
    MPI_Type_vector( 3, 2, M, MPI_INT, &subMat );
    MPI_Type_commit( &subMat );

    // Will only send messages from 0 to 1
    int src = 0;
    int dest = 1;
    int tag = 5;

    // Send the 0th row
    if ( world_rank == 0 ) {
        MPI_Send( &A[0][0], 1, matRow, dest, tag, MPI_COMM_WORLD );
    }
    else if ( world_rank == 1 ) {
        MPI_Recv( &A[0][0], 1, matRow, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE );
    }

    // Let's print the matrix on rank 1 for illustration
    // This also resets the matrix
    if ( world_rank == 1 ) {
        printf( "Should have received the first row:\n" );
        for ( i = 0; i < N; i++ ) {
            for ( j = 0; j < M; j++ ) {
                printf( "%d ", A[i][j] );
                A[i][j] = world_rank;
            }
            printf( "\n" );
        }
        printf( "\n" );
    }

    // Send the 0th column
    if ( world_rank == 0 ) {
        MPI_Send( &A[0][0], 1, matCol, dest, tag, MPI_COMM_WORLD );
    }
    else if ( world_rank == 1 ) {
        MPI_Recv( &A[0][0], 1, matCol, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE );
    }

    // Let's print the matrix on rank 1 for illustration
    // This also resets the matrix
    if ( world_rank == 1 ) {
        printf( "Should have received the 0th column:\n" );
        for ( i = 0; i < N; i++ ) {
            for ( j = 0; j < M; j++ ) {
                printf( "%d ", A[i][j] );
                A[i][j] = world_rank;
            }
            printf( "\n" );
        }
        printf( "\n" );
    }

    // Send the 0th column, but receive into the 4th column
    if ( world_rank == 0 ) {
        MPI_Send( &A[0][0], 1, matCol, dest, tag, MPI_COMM_WORLD );
    }
    else if ( world_rank == 1 ) {
        MPI_Recv( &A[0][4], 1, matCol, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE );
    }

    // Let's print the matrix on rank 1 for illustration
    // This also resets the matrix
    if ( world_rank == 1 ) {
        printf( "Should have received into the 4th column:\n" );
        for ( i = 0; i < N; i++ ) {
            for ( j = 0; j < M; j++ ) {
                printf( "%d ", A[i][j] );
                A[i][j] = world_rank;
            }
            printf( "\n" );
        }
        printf( "\n" );
    }

    // Send a submatrix
    if ( world_rank == 0 ) {
        MPI_Send( &A[0][0], 1, subMat, dest, tag, MPI_COMM_WORLD );
    }
    else if ( world_rank == 1 ) {
        MPI_Recv( &A[0][0], 1, subMat, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE );
    }

    // Let's print the matrix on rank 1 for illustration
    // This also resets the matrix
    if ( world_rank == 1 ) {
        printf( "Should have received into the top left corner:\n" );
        for ( i = 0; i < N; i++ ) {
            for ( j = 0; j < M; j++ ) {
                printf( "%d ", A[i][j] );
                A[i][j] = world_rank;
            }
            printf( "\n" );
        }
        printf( "\n" );
    }
    MPI_Type_free( &matRow );
    MPI_Type_free( &matCol );
    MPI_Type_free( &subMat );
    ierror = MPI_Finalize();
    return ierror;
}

File: Solutions/derived/derived.f90
program dervied
    use mpi
    implicit none

    integer (kind=4), parameter :: N=10,M=10
    integer (kind=4) :: i,j,ierror,world_size,world_rank
    integer (kind=4) :: src,dest,tag
    integer (kind=4), target :: A(N,M)
    integer (kind=4) :: matRow, matCol, subMat
    integer (kind=4), pointer :: ptr


    call MPI_Init( ierror )
    if (ierror .ne. 0) stop ' Cannot initialize MPI '


! Find our world rank & size to print it
    call MPI_Comm_rank( MPI_COMM_WORLD, world_rank, ierror )
    call MPI_Comm_size( MPI_COMM_WORLD, world_size, ierror )

    if ( world_size .le. 1 ) then
        write(6,*) 'Error: must be run on more that 1 process'
        call MPI_Abort( MPI_COMM_WORLD, 10, ierror )
    endif

    call MPI_Type_vector( M, 1, N, MPI_INTEGER, matRow, ierror )
    call MPI_Type_commit( matRow, ierror )
    call MPI_Type_vector( 1, N, 1, MPI_INTEGER, matCol, ierror )
    call MPI_Type_commit( matCol, ierror )
    call MPI_Type_vector( 3, 2, N, MPI_INTEGER, subMat, ierror )
    call MPI_Type_commit( subMat, ierror )

    call MPI_Type_size(subMat,i,ierror)
    write(6,*) ' Size ',i
    call MPI_Type_extent(SubMat,i,ierror)
     write(6,*) ' extent ',i

    do j = 1,M
        do i = 1,N
            A(i,j) = world_rank;
        end do
    end do

! Will only send messages from 0 to 1
    src = 0; dest = 1; tag = 5;

! Send the 0th row
    if ( world_rank .eq. 0 ) then
        ptr => A(1,1)
        call MPI_Send( ptr, 1, matRow, dest, tag, MPI_COMM_WORLD, ierror )

    else if ( world_rank .eq. 1 ) then
        ptr => A(1,1)
        call MPI_Recv( ptr, 1, matRow, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror );
    endif

! Let's print the matrix on rank 1 for illustration
! This also resets the matrix
    if ( world_rank .eq. 1 ) then
        write(6,*) 'Should have received the first row:'
        do i = 1,N
            do j = 1,M
                write(6,'(i4)',advance='NO') A(i,j)
                A(i,j) = world_rank
            end do
            write(6,*)
        end do
        write(6,*)
    endif

! Send the 0th column
    if ( world_rank .eq. 0 ) then
        ptr => A(1,1)
        call MPI_Send( ptr, 1, matCol, dest, tag, MPI_COMM_WORLD, ierror )

    else if ( world_rank .eq. 1 ) then
         ptr => A(1,1)
        call MPI_Recv( ptr, 1, matCol, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror )
    endif

! Let's print the matrix on rank 1 for illustration
! This also resets the matrix
    if ( world_rank .eq. 1 ) then
        write(6,*) 'Should have received the 1st column:'
        do  i = 1,N
            do j = 1,M
                write(6,'(i4)',advance='NO') A(i,j)
                A(i,j) = world_rank
            end do
            write(6,*)
        end do
        write(6,*)
    endif

! Send the 0th column, but receive into the 4th column
    if ( world_rank .eq. 0 ) then
        ptr => A(1,1)
        call MPI_Send( ptr, 1, matCol, dest, tag, MPI_COMM_WORLD, ierror )

    else if ( world_rank .eq. 1 ) then
        ptr => A(1,5)
        call MPI_Recv( ptr, 1, matCol, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror )
    endif

! Let's print the matrix on rank 1 for illustration
! This also resets the matrix
    if ( world_rank .eq. 1 ) then
        write(6,*) 'Should have received into the 5th column:'
        do  i = 1,N
            do j = 1,M
                write(6,'(i4)',advance='NO')  A(i,j)
                A(i,j) = world_rank
            end do
            write(6,*)
        end do
        write(6,*)
    endif

! Send a submatrix
    if ( world_rank .eq. 0 ) then
        ptr => A(1,1)
        call MPI_Send( ptr, 1, subMat, dest, tag, MPI_COMM_WORLD, ierror )

    else if ( world_rank .eq. 1 ) then
        ptr => A(1,1)
        call MPI_Recv( ptr, 1, subMat, src, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierror )
    endif

! Let's print the matrix on rank 1 for illustration
! This also resets the matrix
    if ( world_rank .eq. 1 ) then
        write(6,*) 'Should have received into the top left corner:'
        do i = 1,N 
            do  j = 1,M 
                write(6,'(i4)',advance='NO') A(i,j)
                A(i,j) = world_rank
            end do
            write(6,*)
        end do
        write(6,*)
    endif

    call MPI_Finalize(ierror)
end program dervied
